In [1]:
import pandas as pd

# DATA LOADING

In [13]:
df = pd.read_csv('dataset/Student_Marks.csv')
df.head(10)

number_courses  time_study   Marks
0               3       4.508  19.202
1               4       0.096   7.734
2               4       3.133  13.811
3               6       7.909  53.018
4               8       7.811  55.299
5               6       3.211  17.822
6               3       6.063  29.889
7               5       3.413  17.264
8               4       4.410  20.348
9               3       6.173  30.862

In [14]:
df = pd.read_csv('dataset/ref.csv', sep=';')
df

ch1  ch2  ch3
0   10    1  100
1   20    2  200
2   30    3  300
3   40    4  400
4   50    5  500

In [15]:
df = pd.read_csv('https://raw.githubusercontent.com/ardhiraka/PFDS_sources/master/nbaallelo.csv')
df.head(10)

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   
2          2  194611020CHS   NBA        0     1947  11/2/1946           1   
3          2  194611020CHS   NBA        1     1947  11/2/1946           2   
4          3  194611020DTF   NBA        0     1947  11/2/1946           1   
5          3  194611020DTF   NBA        1     1947  11/2/1946           1   
6          4  194611020PRO   NBA        1     1947  11/2/1946           1   
7          4  194611020PRO   NBA        0     1947  11/2/1946           1   
8          5  194611020STB   NBA        1     1947  11/2/1946           1   
9          5  194611020STB   NBA        0     1947  11/2/1946           1   

   is_playoffs team_id       fran_id  ...   win_equiv  opp_id      opp_fran  \
0            0     TRH       Huskies  ...   40.294830     NYK        Knicks   
1            0     NYK        Knicks  ...   41.705170     TRH       Huskies   
2            0     CHS         Stags  ...   42.012257     NYK        Knicks   
3            0     NYK        Knicks  ...   40.692783     CHS         Stags   
4            0     DTF       Falcons  ...   38.864048     WSC      Capitols   
5            0     WSC      Capitols  ...   43.135952     DTF       Falcons   
6            0     BOS       Celtics  ...   40.459381     PRO  Steamrollers   
7            0     PRO  Steamrollers  ...   41.540619     BOS       Celtics   
8            0     PIT       Ironmen  ...   40.507980     STB       Bombers   
9            0     STB       Bombers  ...   41.492020     PIT       Ironmen   

   opp_pts  opp_elo_i  opp_elo_n  game_location  game_result  forecast notes  
0       68  1300.0000  1306.7233              H            L  0.640065   NaN  
1       66  1300.0000  1293.2767              A            W  0.359935   NaN  
2       47  1306.7233  1297.0712              H            W  0.631101   NaN  
3       63  1300.0000  1309.6521              A            L  0.368899   NaN  
4       50  1300.0000  1320.3811              H            L  0.640065   NaN  
5       33  1300.0000  1279.6189              A            W  0.359935   NaN  
6       59  1300.0000  1305.1542              A            L  0.359935   NaN  
7       53  1300.0000  1294.8458              H            W  0.640065   NaN  
8       56  1300.0000  1304.6908              A            L  0.359935   NaN  
9       51  1300.0000  1295.3092              H            W  0.640065   NaN  

[10 rows x 23 columns]

In [16]:
df.shape

(126314, 23)

In [22]:
pd.set_option('display.max.columns', None)

In [23]:
df.head(5)

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   
2          2  194611020CHS   NBA        0     1947  11/2/1946           1   
3          2  194611020CHS   NBA        1     1947  11/2/1946           2   
4          3  194611020DTF   NBA        0     1947  11/2/1946           1   

   is_playoffs team_id  fran_id  pts      elo_i      elo_n  win_equiv opp_id  \
0            0     TRH  Huskies   66  1300.0000  1293.2767  40.294830    NYK   
1            0     NYK   Knicks   68  1300.0000  1306.7233  41.705170    TRH   
2            0     CHS    Stags   63  1300.0000  1309.6521  42.012257    NYK   
3            0     NYK   Knicks   47  1306.7233  1297.0712  40.692783    CHS   
4            0     DTF  Falcons   33  1300.0000  1279.6189  38.864048    WSC   

   opp_fran  opp_pts  opp_elo_i  opp_elo_n game_location game_result  \
0    Knicks       68  1300.0000  1306.7233             H           L   
1   Huskies       66  1300.0000  1293.2767             A           W   
2    Knicks       47  1306.7233  1297.0712             H           W   
3     Stags       63  1300.0000  1309.6521             A           L   
4  Capitols       50  1300.0000  1320.3811             H           L   

   forecast notes  
0  0.640065   NaN  
1  0.359935   NaN  
2  0.631101   NaN  
3  0.368899   NaN  
4  0.640065   NaN

# DATA OVERVIEW

In [25]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 126314 entries, 0 to 126313
Data columns (total 23 columns):
gameorder        126314 non-null int64
game_id          126314 non-null object
lg_id            126314 non-null object
_iscopy          126314 non-null int64
year_id          126314 non-null int64
date_game        126314 non-null object
seasongame       126314 non-null int64
is_playoffs      126314 non-null int64
team_id          126314 non-null object
fran_id          126314 non-null object
pts              126314 non-null int64
elo_i            126314 non-null float64
elo_n            126314 non-null float64
win_equiv        126314 non-null float64
opp_id           126314 non-null object
opp_fran         126314 non-null object
opp_pts          126314 non-null int64
opp_elo_i        126314 non-null float64
opp_elo_n        126314 non-null float64
game_location    126314 non-null object
game_result      126314 non-null object
forecast         126314 non-null float64
notes     

In [26]:
df.describe().T

count          mean           std          min           25%  \
gameorder    126314.0  31579.000000  18231.927643     1.000000  15790.000000   
_iscopy      126314.0      0.500000      0.500002     0.000000      0.000000   
year_id      126314.0   1988.200374     17.582309  1947.000000   1975.000000   
seasongame   126314.0     43.533733     25.375178     1.000000     22.000000   
is_playoffs  126314.0      0.063857      0.244499     0.000000      0.000000   
pts          126314.0    102.729982     14.814845     0.000000     93.000000   
elo_i        126314.0   1495.236055    112.139945  1091.644500   1417.237975   
elo_n        126314.0   1495.236055    112.461687  1085.774400   1416.994900   
win_equiv    126314.0     41.707889     10.627332    10.152501     34.103035   
opp_pts      126314.0    102.729982     14.814845     0.000000     93.000000   
opp_elo_i    126314.0   1495.236055    112.139945  1091.644500   1417.237975   
opp_elo_n    126314.0   1495.236055    112.461687  1085.774400   1416.994900   
forecast     126314.0      0.500000      0.215252     0.020447      0.327989   

                      50%           75%           max  
gameorder    31579.000000  47368.000000  63157.000000  
_iscopy          0.500000      1.000000      1.000000  
year_id       1990.000000   2003.000000   2015.000000  
seasongame      43.000000     65.000000    108.000000  
is_playoffs      0.000000      0.000000      1.000000  
pts            103.000000    112.000000    186.000000  
elo_i         1500.945550   1576.060000   1853.104500  
elo_n         1500.954400   1576.291625   1853.104500  
win_equiv       42.113357     49.635328     71.112038  
opp_pts        103.000000    112.000000    186.000000  
opp_elo_i     1500.945550   1576.060000   1853.104500  
opp_elo_n     1500.954400   1576.291625   1853.104500  
forecast         0.500000      0.672011      0.979553

In [59]:
df.sample(3)

gameorder       game_id lg_id  _iscopy  year_id   date_game  \
43411       21706  198003190PHO   NBA        1     1980   3/19/1980   
57261       28631  198703250IND   NBA        1     1987   3/25/1987   
123982      61992  201411170BOS   NBA        0     2015  11/17/2014   

        seasongame  is_playoffs team_id  fran_id  pts      elo_i      elo_n  \
43411           77            0     LAL   Lakers  108  1679.8910  1673.9186   
57261           72            0     MIL    Bucks  108  1584.6633  1570.6068   
123982           9            0     BOS  Celtics  114  1391.9214  1386.3912   

        win_equiv opp_id opp_fran  opp_pts  opp_elo_i  opp_elo_n  \
43411   57.784306    PHO     Suns      112  1595.2145  1601.1869   
57261   47.993328    IND   Pacers      125  1495.7330  1509.7896   
123982  28.970966    PHO     Suns      118  1527.1173  1532.6475   

       game_location game_result  forecast notes  
43411              A           L  0.477962   NaN  
57261              A           L  0.484075   NaN  
123982             H           L  0.449522   NaN

In [62]:
pd.DataFrame(df.iloc[57261, :]).T

gameorder       game_id lg_id _iscopy year_id  date_game seasongame  \
57261     28631  198703250IND   NBA       1    1987  3/25/1987         72   

      is_playoffs team_id fran_id  pts    elo_i    elo_n win_equiv opp_id  \
57261           0     MIL   Bucks  108  1584.66  1570.61   47.9933    IND   

      opp_fran opp_pts opp_elo_i opp_elo_n game_location game_result  \
57261   Pacers     125   1495.73   1509.79             A           L   

       forecast notes  
57261  0.484075   NaN

# DATA EXPLORATION

In [27]:
df.head(3)

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   
2          2  194611020CHS   NBA        0     1947  11/2/1946           1   

   is_playoffs team_id  fran_id  pts   elo_i      elo_n  win_equiv opp_id  \
0            0     TRH  Huskies   66  1300.0  1293.2767  40.294830    NYK   
1            0     NYK   Knicks   68  1300.0  1306.7233  41.705170    TRH   
2            0     CHS    Stags   63  1300.0  1309.6521  42.012257    NYK   

  opp_fran  opp_pts  opp_elo_i  opp_elo_n game_location game_result  forecast  \
0   Knicks       68  1300.0000  1306.7233             H           L  0.640065   
1  Huskies       66  1300.0000  1293.2767             A           W  0.359935   
2   Knicks       47  1306.7233  1297.0712             H           W  0.631101   

  notes  
0   NaN  
1   NaN  
2   NaN

In [28]:
df['game_result'].unique()

array(['L', 'W'], dtype=object)

In [30]:
df['fran_id'].value_counts()

Lakers          6024
Celtics         5997
Knicks          5769
Warriors        5657
Pistons         5650
Sixers          5644
Hawks           5572
Kings           5475
Wizards         4582
Spurs           4309
Bulls           4307
Pacers          4227
Thunder         4178
Rockets         4154
Nuggets         4120
Nets            4106
Suns            4080
Bucks           4034
Trailblazers    3870
Cavaliers       3810
Clippers        3733
Jazz            3555
Mavericks       3013
Heat            2371
Pelicans        2254
Magic           2207
Timberwolves    2131
Grizzlies       1657
Raptors         1634
Hornets          894
Colonels         846
Squires          799
Spirits          777
Stars            756
Sounds           697
Baltimore        467
Floridians       440
Condors          430
Capitols         291
Olympians        282
Sails            274
Stags            260
Bombers          249
Steamrollers     168
Packers           72
Redskins          65
Rebels            63
Waterloo     

In [35]:
df.sort_values(by='pts', ascending=True)[['year_id', 'fran_id', 'pts']].head(10)

year_id   fran_id  pts
26684     1973   Nuggets    0
26685     1973   Squires    2
3190      1951    Lakers   18
3191      1951   Pistons   19
4         1947   Falcons   33
320       1947  Capitols   38
628       1947   Huskies   40
129       1947   Ironmen   40
340       1947   Celtics   43
351       1947   Ironmen   43

In [48]:
df.loc[df['fran_id']=='Knicks', :].head(4)

gameorder       game_id lg_id  _iscopy  year_id   date_game  seasongame  \
1           1  194611010TRH   NBA        1     1947   11/1/1946           1   
3           2  194611020CHS   NBA        1     1947   11/2/1946           2   
23         12  194611070STB   NBA        1     1947   11/7/1946           3   
37         19  194611110NYK   NBA        0     1947  11/11/1946           4   

    is_playoffs team_id fran_id  pts      elo_i      elo_n  win_equiv opp_id  \
1             0     NYK  Knicks   68  1300.0000  1306.7233  41.705170    TRH   
3             0     NYK  Knicks   47  1306.7233  1297.0712  40.692783    CHS   
23            0     NYK  Knicks   68  1297.0712  1307.3197  41.767715    STB   
37            0     NYK  Knicks   68  1307.3197  1293.2161  40.288467    CHS   

   opp_fran  opp_pts  opp_elo_i  opp_elo_n game_location game_result  \
1   Huskies       66  1300.0000  1293.2767             A           W   
3     Stags       63  1300.0000  1309.6521             A           L   
23  Bombers       63  1312.8473  1302.5988             A           W   
37    Stags       78  1315.2522  1329.3558             H           L   

    forecast notes  
1   0.359935   NaN  
3   0.368899   NaN  
23  0.339290   NaN  
37  0.629479   NaN

In [40]:
df.loc[df['fran_id']=='Knicks','opp_fran'].value_counts()

Celtics         526
Sixers          471
Pistons         382
Hawks           368
Warriors        335
Wizards         323
Lakers          310
Kings           299
Bulls           251
Pacers          205
Bucks           204
Cavaliers       203
Nets            189
Rockets         145
Heat            137
Thunder         124
Clippers        122
Suns            120
Trailblazers    105
Magic            99
Spurs            90
Jazz             90
Pelicans         90
Raptors          85
Nuggets          81
Mavericks        68
Baltimore        62
Timberwolves     50
Hornets          40
Grizzlies        37
Capitols         32
Bombers          23
Stags            23
Steamrollers     20
Olympians        20
Rebels            9
Huskies           6
Falcons           6
Ironmen           6
Jets              5
Redskins          2
Denver            2
Packers           2
Waterloo          2
Name: opp_fran, dtype: int64

In [55]:
df.loc[df['fran_id']=='Huskies', 'year_id'].agg(['max', 'min'])

max    1947
min    1947
Name: year_id, dtype: int64

tim huskies hanya bermian di tahun 1947

In [56]:
df.loc[df['fran_id']=='Knicks', 'year_id'].agg(['max', 'min'])

max    2015
min    1947
Name: year_id, dtype: int64

tim knicks bermain dari season 1947 hingga 2015

In [57]:
df.loc[df['fran_id']=='Warriors', 'pts'].sum()

591224

In [58]:
df.loc[df['fran_id']=='Warriors', ['seasongame', 'pts']].sum()

seasongame    237531
pts           591224
dtype: int64

tim warriors mengumpulkan poin 591224 sepanjang masa

# QUERY DATASET

In [63]:
df.sample(5)

gameorder       game_id lg_id  _iscopy  year_id   date_game  \
98183       49092  200411220UTA   NBA        0     2005  11/22/2004   
73604       36803  199405020IND   NBA        0     1994    5/2/1994   
107238      53620  200802080TOR   NBA        1     2008    2/8/2008   
18073        9037  196901050SFW   NBA        0     1969    1/5/1969   
46590       23296  198201270DET   NBA        0     1982   1/27/1982   

        seasongame  is_playoffs team_id   fran_id  pts      elo_i      elo_n  \
98183           11            0     UTA      Jazz   75  1557.3202  1549.5525   
73604           85            1     IND    Pacers   99  1613.3756  1621.3735   
107238          47            0     LAC  Clippers  102  1405.3348  1418.5658   
18073           42            0     SFW  Warriors   86  1400.3191  1375.8175   
46590           44            0     DET   Pistons  108  1469.1750  1471.9307   

        win_equiv opp_id  opp_fran  opp_pts  opp_elo_i  opp_elo_n  \
98183   46.498756    NOH  Pelicans       76  1430.6466  1438.4143   
73604   53.680355    ORL     Magic       86  1608.6635  1600.6655   
107238  32.424175    TOR   Raptors       98  1572.0817  1558.8507   
18073   33.933945    BOS   Celtics      134  1560.3168  1584.8184   
46590   38.295700    ATL     Hawks      107  1474.2969  1471.5413   

       game_location game_result  forecast notes  
98183              H           L  0.786648   NaN  
73604              H           W  0.646290   NaN  
107238             A           W  0.177188   NaN  
18073              H           L  0.414505   NaN  
46590              H           W  0.633245   NaN

In [70]:
df[
    (df['year_id']==2001) &
    (df['fran_id']=='Celtics') &
    (df['game_result']=='W') &
    (df['game_location']=='H')
]

gameorder       game_id lg_id  _iscopy  year_id   date_game  \
87776      43889  200011010BOS   NBA        0     2001   11/1/2000   
87805      43903  200011030BOS   NBA        0     2001   11/3/2000   
87970      43986  200011150BOS   NBA        0     2001  11/15/2000   
88001      44001  200011170BOS   NBA        0     2001  11/17/2000   
88068      44035  200011220BOS   NBA        0     2001  11/22/2000   
88093      44047  200011240BOS   NBA        0     2001  11/24/2000   
88233      44117  200012030BOS   NBA        0     2001   12/3/2000   
88336      44169  200012100BOS   NBA        0     2001  12/10/2000   
88469      44235  200012200BOS   NBA        0     2001  12/20/2000   
88748      44375  200101100BOS   NBA        0     2001   1/10/2001   
88925      44463  200101230BOS   NBA        0     2001   1/23/2001   
88975      44488  200101260BOS   NBA        0     2001   1/26/2001   
89013      44507  200101290BOS   NBA        0     2001   1/29/2001   
89105      44553  200102040BOS   NBA        0     2001    2/4/2001   
89141      44571  200102070BOS   NBA        0     2001    2/7/2001   
89364      44683  200102260BOS   NBA        0     2001   2/26/2001   
89425      44713  200103020BOS   NBA        0     2001    3/2/2001   
89525      44763  200103090BOS   NBA        0     2001    3/9/2001   
89756      44879  200103250BOS   NBA        0     2001   3/25/2001   
90057      45029  200104150BOS   NBA        0     2001   4/15/2001   

       seasongame  is_playoffs team_id  fran_id  pts      elo_i      elo_n  \
87776           1            0     BOS  Celtics  103  1496.9196  1509.4957   
87805           2            0     BOS  Celtics   93  1509.4957  1512.2125   
87970           7            0     BOS  Celtics  116  1483.8416  1487.9716   
88001           8            0     BOS  Celtics  106  1487.9716  1490.5791   
88068          11            0     BOS  Celtics   96  1460.4922  1472.5193   
88093          12            0     BOS  Celtics  103  1472.5193  1477.3024   
88233          17            0     BOS  Celtics   96  1464.1198  1466.8375   
88336          20            0     BOS  Celtics  104  1450.2649  1453.4750   
88469          25            0     BOS  Celtics   99  1397.2222  1405.6613   
88748          36            0     BOS  Celtics   88  1368.3331  1384.8729   
88925          42            0     BOS  Celtics   83  1384.1893  1386.8386   
88975          43            0     BOS  Celtics   99  1386.8386  1404.1323   
89013          44            0     BOS  Celtics   99  1404.1323  1412.0132   
89105          47            0     BOS  Celtics  103  1437.8401  1448.2347   
89141          49            0     BOS  Celtics   94  1442.6997  1449.4445   
89364          57            0     BOS  Celtics   85  1468.3358  1472.3491   
89425          59            0     BOS  Celtics  103  1468.3030  1476.5731   
89525          62            0     BOS  Celtics  104  1451.1699  1459.3822   
89756          69            0     BOS  Celtics  102  1477.7058  1479.0504   
90057          81            0     BOS  Celtics   95  1463.9949  1467.3157   

       win_equiv opp_id   opp_fran  opp_pts  opp_elo_i  opp_elo_n  \
87776  41.935509    DET    Pistons       83  1529.7430  1517.1669   
87805  42.220291    TOR    Raptors       91  1472.1443  1469.4275   
87970  39.678341    WAS    Wizards      109  1418.0574  1413.9274   
88001  39.951633    ATL      Hawks       99  1332.2992  1329.6917   
88068  38.062054    HOU    Rockets       81  1530.7407  1518.7136   
88093  38.561615    ORL      Magic       98  1476.9380  1472.1549   
88233  37.469780    LAC   Clippers       87  1289.0637  1286.3459   
88336  36.083122    DEN    Nuggets      102  1449.5248  1446.3147   
88469  31.235086    NJN       Nets       86  1404.6704  1396.2313   
88748  29.205793    MIA       Heat       76  1575.6382  1559.0984   
88925  29.395197    ATL      Hawks       82  1380.4136  1377.7643   
88975  31.083946    PHO       Suns       83  1546.8582  1529.5645   


In [71]:
df[
    (df['year_id']==2001) &
    (df['fran_id']=='Celtics') &
    (df['game_result']=='W') &
    (df['game_location']=='H')
][['opp_fran', 'opp_pts']]

opp_fran  opp_pts
87776    Pistons       83
87805    Raptors       91
87970    Wizards      109
88001      Hawks       99
88068    Rockets       81
88093      Magic       98
88233   Clippers       87
88336    Nuggets      102
88469       Nets       86
88748       Heat       76
88925      Hawks       82
88975       Suns       83
89013    Pistons       87
89105  Cavaliers       82
89141      Bucks       91
89364    Thunder       82
89425       Jazz       99
89525    Wizards       72
89756      Bulls       98
90057       Nets       88

In [78]:
df.query("year_id==2001 & \
            fran_id=='Celtics' & \
            game_result=='W' & \
            game_location=='H'"
        )

gameorder       game_id lg_id  _iscopy  year_id   date_game  \
87776      43889  200011010BOS   NBA        0     2001   11/1/2000   
87805      43903  200011030BOS   NBA        0     2001   11/3/2000   
87970      43986  200011150BOS   NBA        0     2001  11/15/2000   
88001      44001  200011170BOS   NBA        0     2001  11/17/2000   
88068      44035  200011220BOS   NBA        0     2001  11/22/2000   
88093      44047  200011240BOS   NBA        0     2001  11/24/2000   
88233      44117  200012030BOS   NBA        0     2001   12/3/2000   
88336      44169  200012100BOS   NBA        0     2001  12/10/2000   
88469      44235  200012200BOS   NBA        0     2001  12/20/2000   
88748      44375  200101100BOS   NBA        0     2001   1/10/2001   
88925      44463  200101230BOS   NBA        0     2001   1/23/2001   
88975      44488  200101260BOS   NBA        0     2001   1/26/2001   
89013      44507  200101290BOS   NBA        0     2001   1/29/2001   
89105      44553  200102040BOS   NBA        0     2001    2/4/2001   
89141      44571  200102070BOS   NBA        0     2001    2/7/2001   
89364      44683  200102260BOS   NBA        0     2001   2/26/2001   
89425      44713  200103020BOS   NBA        0     2001    3/2/2001   
89525      44763  200103090BOS   NBA        0     2001    3/9/2001   
89756      44879  200103250BOS   NBA        0     2001   3/25/2001   
90057      45029  200104150BOS   NBA        0     2001   4/15/2001   

       seasongame  is_playoffs team_id  fran_id  pts      elo_i      elo_n  \
87776           1            0     BOS  Celtics  103  1496.9196  1509.4957   
87805           2            0     BOS  Celtics   93  1509.4957  1512.2125   
87970           7            0     BOS  Celtics  116  1483.8416  1487.9716   
88001           8            0     BOS  Celtics  106  1487.9716  1490.5791   
88068          11            0     BOS  Celtics   96  1460.4922  1472.5193   
88093          12            0     BOS  Celtics  103  1472.5193  1477.3024   
88233          17            0     BOS  Celtics   96  1464.1198  1466.8375   
88336          20            0     BOS  Celtics  104  1450.2649  1453.4750   
88469          25            0     BOS  Celtics   99  1397.2222  1405.6613   
88748          36            0     BOS  Celtics   88  1368.3331  1384.8729   
88925          42            0     BOS  Celtics   83  1384.1893  1386.8386   
88975          43            0     BOS  Celtics   99  1386.8386  1404.1323   
89013          44            0     BOS  Celtics   99  1404.1323  1412.0132   
89105          47            0     BOS  Celtics  103  1437.8401  1448.2347   
89141          49            0     BOS  Celtics   94  1442.6997  1449.4445   
89364          57            0     BOS  Celtics   85  1468.3358  1472.3491   
89425          59            0     BOS  Celtics  103  1468.3030  1476.5731   
89525          62            0     BOS  Celtics  104  1451.1699  1459.3822   
89756          69            0     BOS  Celtics  102  1477.7058  1479.0504   
90057          81            0     BOS  Celtics   95  1463.9949  1467.3157   

       win_equiv opp_id   opp_fran  opp_pts  opp_elo_i  opp_elo_n  \
87776  41.935509    DET    Pistons       83  1529.7430  1517.1669   
87805  42.220291    TOR    Raptors       91  1472.1443  1469.4275   
87970  39.678341    WAS    Wizards      109  1418.0574  1413.9274   
88001  39.951633    ATL      Hawks       99  1332.2992  1329.6917   
88068  38.062054    HOU    Rockets       81  1530.7407  1518.7136   
88093  38.561615    ORL      Magic       98  1476.9380  1472.1549   
88233  37.469780    LAC   Clippers       87  1289.0637  1286.3459   
88336  36.083122    DEN    Nuggets      102  1449.5248  1446.3147   
88469  31.235086    NJN       Nets       86  1404.6704  1396.2313   
88748  29.205793    MIA       Heat       76  1575.6382  1559.0984   
88925  29.395197    ATL      Hawks       82  1380.4136  1377.7643   
88975  31.083946    PHO       Suns       83  1546.8582  1529.5645   


# GROUPING AND AGGREGATING

In [79]:
df.sample(4)

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
28764      14383  197304030DNR   ABA        0     1973   4/3/1973          87   
92561      46281  200204120TOR   NBA        1     2002  4/12/2002          79   
48663      24332  198302020CLE   NBA        0     1983   2/2/1983          46   
72651      36326  199402160LAC   NBA        1     1994  2/16/1994          50   

       is_playoffs team_id    fran_id  pts      elo_i      elo_n  win_equiv  \
28764            1     DNR    Nuggets  105  1513.1598  1522.5474  47.189938   
92561            0     ATL      Hawks   73  1442.7230  1427.5809  33.318954   
48663            0     CLE  Cavaliers  105  1296.1986  1290.8342  20.973808   
72651            0     UTA       Jazz  103  1555.1611  1560.1274  47.626511   

      opp_id  opp_fran  opp_pts  opp_elo_i  opp_elo_n game_location  \
28764    INA    Pacers       94  1571.5793  1562.1918             H   
92561    TOR   Raptors      112  1476.7701  1491.9122             A   
48663    NJN      Nets      118  1585.1963  1590.5607             H   
72651    LAC  Clippers       99  1393.3961  1388.4298             A   

      game_result  forecast notes  
28764           W  0.559555   NaN  
92561           L  0.316125   NaN  
48663           L  0.252003   NaN  
72651           W  0.587962   NaN

In [81]:
gr = df.groupby('game_result')
gr

In [84]:
gr.get_group('L').head(6)

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0           1  194611010TRH   NBA        0     1947  11/1/1946           1   
3           2  194611020CHS   NBA        1     1947  11/2/1946           2   
4           3  194611020DTF   NBA        0     1947  11/2/1946           1   
6           4  194611020PRO   NBA        1     1947  11/2/1946           1   
8           5  194611020STB   NBA        1     1947  11/2/1946           1   
10          6  194611030CLR   NBA        1     1947  11/3/1946           2   

    is_playoffs team_id  fran_id  pts      elo_i      elo_n  win_equiv opp_id  \
0             0     TRH  Huskies   66  1300.0000  1293.2767  40.294830    NYK   
3             0     NYK   Knicks   47  1306.7233  1297.0712  40.692783    CHS   
4             0     DTF  Falcons   33  1300.0000  1279.6189  38.864048    WSC   
6             0     BOS  Celtics   53  1300.0000  1294.8458  40.459381    PRO   
8             0     PIT  Ironmen   51  1300.0000  1295.3092  40.507980    STB   
10            0     TRH  Huskies   60  1293.2767  1286.1534  39.548164    CLR   

        opp_fran  opp_pts  opp_elo_i  opp_elo_n game_location game_result  \
0         Knicks       68     1300.0  1306.7233             H           L   
3          Stags       63     1300.0  1309.6521             A           L   
4       Capitols       50     1300.0  1320.3811             H           L   
6   Steamrollers       59     1300.0  1305.1542             A           L   
8        Bombers       56     1300.0  1304.6908             A           L   
10        Rebels       71     1300.0  1307.1233             A           L   

    forecast notes  
0   0.640065   NaN  
3   0.368899   NaN  
4   0.640065   NaN  
6   0.359935   NaN  
8   0.359935   NaN  
10  0.351068   NaN

In [85]:
gr.get_group('W').head(6)

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
1           1  194611010TRH   NBA        1     1947  11/1/1946           1   
2           2  194611020CHS   NBA        0     1947  11/2/1946           1   
5           3  194611020DTF   NBA        1     1947  11/2/1946           1   
7           4  194611020PRO   NBA        0     1947  11/2/1946           1   
9           5  194611020STB   NBA        0     1947  11/2/1946           1   
11          6  194611030CLR   NBA        0     1947  11/3/1946           1   

    is_playoffs team_id       fran_id  pts   elo_i      elo_n  win_equiv  \
1             0     NYK        Knicks   68  1300.0  1306.7233  41.705170   
2             0     CHS         Stags   63  1300.0  1309.6521  42.012257   
5             0     WSC      Capitols   50  1300.0  1320.3811  43.135952   
7             0     PRO  Steamrollers   59  1300.0  1305.1542  41.540619   
9             0     STB       Bombers   56  1300.0  1304.6908  41.492020   
11            0     CLR        Rebels   71  1300.0  1307.1233  41.747120   

   opp_id opp_fran  opp_pts  opp_elo_i  opp_elo_n game_location game_result  \
1     TRH  Huskies       66  1300.0000  1293.2767             A           W   
2     NYK   Knicks       47  1306.7233  1297.0712             H           W   
5     DTF  Falcons       33  1300.0000  1279.6189             A           W   
7     BOS  Celtics       53  1300.0000  1294.8458             H           W   
9     PIT  Ironmen       51  1300.0000  1295.3092             H           W   
11    TRH  Huskies       60  1293.2767  1286.1534             H           W   

    forecast notes  
1   0.359935   NaN  
2   0.631101   NaN  
5   0.359935   NaN  
7   0.640065   NaN  
9   0.640065   NaN  
11  0.648932   NaN

In [90]:
df.groupby('game_result')['pts'].mean()

game_result
L     97.349209
W    108.110756
Name: pts, dtype: float64

grup yg kalah memiliki agregat rata2 97, sedangkan yg menang memiliki agregat rata2 108

In [95]:
df[
    (df['year_id']>2010) &
    (df['fran_id']=='Spurs')
].groupby(['game_result', 'year_id'])['game_id'].count()

game_result  year_id
L            2011       25
             2012       20
             2013       30
             2014       27
             2015       31
W            2011       63
             2012       60
             2013       73
             2014       78
             2015       58
Name: game_id, dtype: int64

In [96]:
df[
    (df['year_id']>2010) &
    (df['fran_id']=='Spurs')
].groupby(['game_result', 'year_id']).agg({'pts': 'mean', 'game_id': 'count'})

pts  game_id
game_result year_id                     
L           2011      93.840000       25
            2012      94.350000       20
            2013      93.333333       30
            2014      96.925926       27
            2015      97.548387       31
W           2011     106.698413       63
            2012     106.366667       60
            2013     106.232877       73
            2014     108.564103       78
            2015     106.224138       58

# MANIPULATING COLUMNS

In [97]:
df.sample(4)

gameorder       game_id lg_id  _iscopy  year_id   date_game  \
95721       47861  200311240BOS   NBA        0     2004  11/24/2003   
43660       21831  198010110ATL   NBA        0     1981  10/11/1980   
37730       18866  197704030NYK   NBA        1     1977    4/3/1977   
104290      52146  200701170MEM   NBA        1     2007   1/17/2007   

        seasongame  is_playoffs team_id   fran_id  pts      elo_i      elo_n  \
95721           13            0     BOS   Celtics   88  1467.5905  1456.0553   
43660            1            0     ATL     Hawks  101  1542.9641  1546.8525   
37730           78            0     BUF  Clippers  111  1396.9695  1391.1243   
104290          39            0     PHI    Sixers  102  1387.0313  1378.0940   

        win_equiv opp_id   opp_fran  opp_pts  opp_elo_i  opp_elo_n  \
95721   36.085445    NYK     Knicks       94  1443.2512  1454.7864   
43660   46.373188    CHI      Bulls       93  1449.5720  1445.6836   
37730   29.622879    NYK     Knicks      130  1519.1339  1524.9791   
104290  28.574066    MEM  Grizzlies      118  1397.6432  1406.5804   

       game_location game_result  forecast notes  
95721              H           L  0.671671   NaN  
43660              H           W  0.752735   NaN  
37730              A           L  0.217741   NaN  
104290             A           L  0.345985   NaN

In [99]:
#create a new column
df['selisih_poin'] = df['pts'] - df['opp_pts']
df.head(5)

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   
2          2  194611020CHS   NBA        0     1947  11/2/1946           1   
3          2  194611020CHS   NBA        1     1947  11/2/1946           2   
4          3  194611020DTF   NBA        0     1947  11/2/1946           1   

   is_playoffs team_id  fran_id  pts      elo_i      elo_n  win_equiv opp_id  \
0            0     TRH  Huskies   66  1300.0000  1293.2767  40.294830    NYK   
1            0     NYK   Knicks   68  1300.0000  1306.7233  41.705170    TRH   
2            0     CHS    Stags   63  1300.0000  1309.6521  42.012257    NYK   
3            0     NYK   Knicks   47  1306.7233  1297.0712  40.692783    CHS   
4            0     DTF  Falcons   33  1300.0000  1279.6189  38.864048    WSC   

   opp_fran  opp_pts  opp_elo_i  opp_elo_n game_location game_result  \
0    Knicks       68  1300.0000  1306.7233             H           L   
1   Huskies       66  1300.0000  1293.2767             A           W   
2    Knicks       47  1306.7233  1297.0712             H           W   
3     Stags       63  1300.0000  1309.6521             A           L   
4  Capitols       50  1300.0000  1320.3811             H           L   

   forecast notes  selisih_poin  
0  0.640065   NaN            -2  
1  0.359935   NaN             2  
2  0.631101   NaN            16  
3  0.368899   NaN           -16  
4  0.640065   NaN           -17

In [101]:
nba=df.copy()

In [104]:
#rename column
nba.rename(columns={'game_location': 'location', 'game_result': 'result'}, inplace=True)

In [105]:
nba.head()

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   
2          2  194611020CHS   NBA        0     1947  11/2/1946           1   
3          2  194611020CHS   NBA        1     1947  11/2/1946           2   
4          3  194611020DTF   NBA        0     1947  11/2/1946           1   

   is_playoffs team_id  fran_id  pts      elo_i      elo_n  win_equiv opp_id  \
0            0     TRH  Huskies   66  1300.0000  1293.2767  40.294830    NYK   
1            0     NYK   Knicks   68  1300.0000  1306.7233  41.705170    TRH   
2            0     CHS    Stags   63  1300.0000  1309.6521  42.012257    NYK   
3            0     NYK   Knicks   47  1306.7233  1297.0712  40.692783    CHS   
4            0     DTF  Falcons   33  1300.0000  1279.6189  38.864048    WSC   

   opp_fran  opp_pts  opp_elo_i  opp_elo_n location result  forecast notes  \
0    Knicks       68  1300.0000  1306.7233        H      L  0.640065   NaN   
1   Huskies       66  1300.0000  1293.2767        A      W  0.359935   NaN   
2    Knicks       47  1306.7233  1297.0712        H      W  0.631101   NaN   
3     Stags       63  1300.0000  1309.6521        A      L  0.368899   NaN   
4  Capitols       50  1300.0000  1320.3811        H      L  0.640065   NaN   

   selisih_poin  
0            -2  
1             2  
2            16  
3           -16  
4           -17

In [107]:
#drop columns
nba.drop(['elo_i', 'elo_n'], axis=1, inplace=True)

In [108]:
nba.head()

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   
2          2  194611020CHS   NBA        0     1947  11/2/1946           1   
3          2  194611020CHS   NBA        1     1947  11/2/1946           2   
4          3  194611020DTF   NBA        0     1947  11/2/1946           1   

   is_playoffs team_id  fran_id  pts  win_equiv opp_id  opp_fran  opp_pts  \
0            0     TRH  Huskies   66  40.294830    NYK    Knicks       68   
1            0     NYK   Knicks   68  41.705170    TRH   Huskies       66   
2            0     CHS    Stags   63  42.012257    NYK    Knicks       47   
3            0     NYK   Knicks   47  40.692783    CHS     Stags       63   
4            0     DTF  Falcons   33  38.864048    WSC  Capitols       50   

   opp_elo_i  opp_elo_n location result  forecast notes  selisih_poin  
0  1300.0000  1306.7233        H      L  0.640065   NaN            -2  
1  1300.0000  1293.2767        A      W  0.359935   NaN             2  
2  1306.7233  1297.0712        H      W  0.631101   NaN            16  
3  1300.0000  1309.6521        A      L  0.368899   NaN           -16  
4  1300.0000  1320.3811        H      L  0.640065   NaN           -17